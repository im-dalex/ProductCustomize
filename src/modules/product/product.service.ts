import { api } from "@/core/api";
import { Product } from "@/modules/product/models/product.interface";

const CONTROLLER = "api/Product";

export const getProducts = (
  from: number,
  to: number,
  searchString?: string
): Promise<Product[]> =>
  api.get(`${CONTROLLER}/GetProductsFromTo`, {
    params: { searchString, from, to },
  });
