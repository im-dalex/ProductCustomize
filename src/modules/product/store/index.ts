import { defineStore } from 'pinia';
import { ProductState } from '@/modules/product/models/product.interface';

export const useProductStore = defineStore('product', {
  persist: true,
  state: () => ({
    selectedProduct: {} as ProductState,
  }),
});
