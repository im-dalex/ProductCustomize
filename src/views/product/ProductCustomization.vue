<template>
  <div class="row mx-0 row-cols-1 row-cols-md-2 g-4 gx-md-5">
    <div class="col">
      <canvas id="product-preview"></canvas>
    </div>
    <div class="col">
      <div class="product-customize">
        <div class="product-info">
          <span class="fw-bold">{{ selectedProduct.name }}</span>
          <span>{{ selectedProduct.createdDate }}</span>
        </div>
        <dynamic-form
          class="mt-4"
          :fields="fields"
          @text-change="setText"
          @font-change="setFont"
        />
      </div>
    </div>
    <div class="col">
      <button type="button" class="btn btn-secondary rounded-0" @click="goBack">
        Back
      </button>
    </div>
  </div>
</template>

<script lang="ts">
import { defineComponent, PropType } from 'vue';
import { mapState } from 'pinia';
import {
  DropdownOption,
  FieldUpdate,
  ProductField,
} from '@/modules/product/models/product.interface';
import { useProductStore } from '@/modules/product/store';
import DynamicForm from '@/views/product/components/DynamicProductForm.vue';

export default defineComponent({
  name: 'ProductCustomization',
  components: { DynamicForm },
  props: {
    productId: {
      type: String as PropType<string>,
      required: true,
    },
  },
  data: () => ({
    fields: [] as ProductField[],
  }),
  computed: {
    ...mapState(useProductStore, ['selectedProduct']),
    customilyLib() {
      return window.engraver;
    },
  },
  async mounted() {
    await this.initProductCanvas();
  },
  methods: {
    async initProductCanvas(): Promise<void> {
      this.customilyLib.init('product-preview');
      await this.customilyLib.setProduct(this.productId);
      const { preview: product } = this.customilyLib.currentProduct;

      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      this.fields = product.textsPreview.map((field: any) => ({
        id: field.id,
        fontList: this.mapDropdownOptions(field.fontsMap),
        currentFont: field.fontPath,
      }));
    },
    mapDropdownOptions(strOptions: string): DropdownOption[] {
      const options = JSON.parse(strOptions);
      return options.map((option: [number, string]) => ({
        id: option[0],
        value: option[1],
        label: this.getFontLabel(option[1]),
      }));
    },
    getFontLabel(font: string): string | undefined {
      const fontType = font.split('/').pop();
      return fontType?.split('-').slice(0, 2).join(' ');
    },
    setText(option: FieldUpdate): void {
      this.customilyLib.setText(option.fieldId, option.value);
    },
    setFont(option: FieldUpdate): void {
      this.customilyLib.setFont(option.fieldId, option.value);
    },
    goBack() {
      this.$router.go(-1);
    },
  },
});
</script>

<style lang="scss" scoped>
.product-customize {
  padding: 1rem;
  height: 100%;
  background: #ebebeb;
  .product-info {
    display: flex;
    align-items: center;
    justify-content: space-between;
  }
}
.btn-secondary {
  width: 40%;
}
@media only screen and (max-width: 768px) {
  .btn-secondary {
    width: 100%;
  }
}
</style>
