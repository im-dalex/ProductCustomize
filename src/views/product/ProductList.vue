<template>
  <div class="product-list row row-cols-1 row-cols-md-2 row-cols-lg-4 mx-0 g-4">
    <div v-for="(product, idx) in products" :key="product.id" class="col">
      <product-card :product="product" @try-it="tryItHandler($event, idx)" />
    </div>
  </div>
  <mq-responsive>
    <template #md->
      <infinite-scroll
        v-model:config="pagination"
        @bottom-reached="getProducts"
      />
    </template>
    <template #lg+>
      <pagination
        v-if="products.length"
        class="d-flex justify-content-center mt-4"
        v-model:config="pagination"
        @page-changed="getProducts"
      />
    </template>
  </mq-responsive>
</template>

<script lang="ts">
import { defineComponent } from 'vue';
import { mapWritableState } from 'pinia';
import { MqResponsive } from 'vue3-mq';
import { useProductStore } from '@/modules/product/store';
import { Product } from '@/modules/product/models/product.interface';
import { getProducts as _getProducts } from '@/modules/product/product.service';
import ProductCard from '@/views/product/components/ProductCard.vue';
import Pagination from '@/components/shared/pagination/Paginator.vue';
import { Paginator } from '@/components/shared/pagination/paginator.interface';
import InfiniteScroll from '@/components/shared/pagination/InfiniteScroll.vue';
import { formatDate } from '@/plugins/filters';

const DEFAULT_PAGE_SIZE = 8;

export default defineComponent({
  name: 'ProductList',
  components: { MqResponsive, ProductCard, Pagination, InfiniteScroll },
  inject: ['mq'],
  data: () => ({
    products: [] as Product[],
    pagination: {
      pageSize: DEFAULT_PAGE_SIZE,
      toItem: DEFAULT_PAGE_SIZE,
      pageCount: 0,
      fromItem: 0,
    } as Paginator,
  }),
  computed: {
    ...mapWritableState(useProductStore, ['selectedProduct']),
  },
  watch: {
    'mq.lgPlus'(newVal, oldVal) {
      if (newVal != oldVal) {
        this.resetPagination();
        this.getProducts();
      }
    },
  },
  async created() {
    await this.getProducts();
  },
  methods: {
    async getProducts(): Promise<void> {
      const data = await _getProducts(
        this.pagination.fromItem,
        this.pagination.toItem
      );
      this.pagination.nextDisabled = data.length == 0;
      this.pagination.pageCount = data.length;
      if (!this.pagination.nextDisabled) {
        this.products = this.mq.lgPlus
          ? [...data]
          : [...this.products, ...data];
      }
    },
    tryItHandler(productId: string, productIdx: number): void {
      const product = this.products[productIdx];
      this.selectedProduct = {
        id: productId,
        name: product.name,
        createdDate: formatDate(product.createdDate),
      };

      this.$router.push({
        name: 'ProductCustomization',
        params: { productId },
      });
    },
    resetPagination() {
      this.products = [];
      this.pagination = {
        pageSize: DEFAULT_PAGE_SIZE,
        toItem: DEFAULT_PAGE_SIZE,
        pageCount: 0,
        fromItem: 0,
      };
    },
  },
});
</script>
