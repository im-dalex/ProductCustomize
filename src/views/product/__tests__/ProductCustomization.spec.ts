import { shallowMount } from '@vue/test-utils';
import { createTestingPinia } from '@pinia/testing';
import flushPromises from 'flush-promises';
import { Wrapper } from 'tests/utils/VueWrapper';
import ProductCustomization from '@/views/product/ProductCustomization.vue';
import DynamicForm from '@/views/product/components/DynamicProductForm.vue';
import { FieldUpdate } from '@/modules/product/models/product.interface';

const mockRouter = { go: jest.fn() };
const mockProduct = 'bfba44e2-9253-48a7-b372-25b4b3a4230e';
const mockEngraver = {
  init: jest.fn(),
  setProduct: jest.fn(),
  setText: jest.fn(),
  setFont: jest.fn(),
  currentProduct: {
    preview: {
      textsPreview: [
        {
          id: 1,
          fontPath:
            '/Content/fonts/frontEndDev/Crimson-Roman-36907fd5-1391-43a1-af38-1007f8bcd689.otf',
          fontsMap:
            '[[1,"/Content/fonts/frontEndDev/Crimson-Roman-36907fd5-1391-43a1-af38-1007f8bcd689.otf"],[2,"/Content/fonts/frontEndDev/AlexBrush-Regular-8ea01138-9fbe-4044-b421-878c0609b083.ttf"]]',
        },
      ],
    },
  },
};

describe('ProductCustomization', () => {
  let mountFunc: Wrapper;
  beforeEach(() => {
    window.engraver = { ...mockEngraver };
    mountFunc = (options) => {
      return shallowMount(ProductCustomization, {
        ...options,
        props: { productId: mockProduct },
        global: {
          ...options?.global,
          plugins: [createTestingPinia()],
        },
      });
    };
  });

  it('gets the product fields', async () => {
    const wrapper = mountFunc();
    await flushPromises();

    const fields = wrapper.vm.fields;
    const formComponent = wrapper.findComponent(DynamicForm);

    expect(fields.length).toBeGreaterThan(0);
    expect(formComponent.exists()).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('returns to the product page', () => {
    const wrapper = mountFunc({
      global: {
        mocks: {
          $router: mockRouter,
        },
      },
    });

    wrapper.find('button').trigger('click');

    expect(mockRouter.go).toHaveBeenCalledTimes(1);
    expect(mockRouter.go).toHaveBeenCalledWith(-1);
  });

  it('updates the font', () => {
    const wrapper = mountFunc();
    const mockOption: FieldUpdate = {
      fieldId: 1,
      value: 1,
    };

    wrapper.vm.setFont(mockOption);
    const setFontFunc = wrapper.vm.customilyLib.setFont;

    expect(setFontFunc).toHaveBeenCalled();
    expect(setFontFunc).toHaveBeenCalledWith(
      mockOption.fieldId,
      mockOption.value
    );
  });

  it('updates the text', () => {
    const wrapper = mountFunc();
    const mockOption: FieldUpdate = {
      fieldId: 1,
      value: 'my text',
    };

    wrapper.vm.setText(mockOption);
    const setTextFunc = wrapper.vm.customilyLib.setText;

    expect(setTextFunc).toHaveBeenCalled();
    expect(setTextFunc).toHaveBeenCalledWith(
      mockOption.fieldId,
      mockOption.value
    );
  });
});
