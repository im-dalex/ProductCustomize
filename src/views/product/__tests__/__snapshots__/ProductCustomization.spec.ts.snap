// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCustomization gets the product fields 1`] = `
<div class="row mx-0 row-cols-1 row-cols-md-2 g-4 gx-md-5">
  <div class="col"><canvas id="product-preview"></canvas></div>
  <div class="col">
    <div class="product-customize">
      <div class="product-info"><span class="fw-bold"></span><span></span></div>
      <dynamic-form-stub fields="[object Object]" class="mt-4"></dynamic-form-stub>
    </div>
  </div>
  <div class="col"><button type="button" class="btn btn-secondary rounded-0"> Back </button></div>
</div>
`;
